src/snippets/snippets.ts

```diff
@@ -1,58 +1,73 @@
 
-export interface RawSnippet {
-	trigger: string | RegExp;
-	replacement: string;
-	options: string;
+/**
+ * in visual snippets, if the replacement is a string, this is the magic substring to indicate the selection.
+ */
+export const VISUAL_SNIPPET_MAGIC_SELECTION_PLACEHOLDER = "${VISUAL}";
+
+/**
+ * a snippet object contains all the information necessary to run a snippet
+ */
+export type Snippet =
+	| VisualSnippet
+	| RegexSnippet
+	| StringSnippet
+
+/**
+ * there are 3 distinct types of snippets:
+ * 
+ * `visual` snippets only trigger on text selections.
+ * visual snippets support only (single-character) string triggers, and string or function replacements.
+ * visual replacement functions take in the text selection and return a string, or `false` to indicate to actually not do anything.
+ * 
+ * `regex` snippets support string (with the "r" raw option set) or regex triggers, and string or function replacements.
+ * regex replacement functions take in the regex match and return a string.
+ * 
+ * `string` snippets support string triggers (when no "r" raw option set), and string or function replacements.
+ * string replacement functions take in the matched string and return a string.
+ */
+export type SnippetType = Snippet["type"];
+
+interface _Snippet {
+	type: string;
+	options: Options;
 	priority?: number;
 	description?: string;
-	flags?: string;
 }
 
-export class ParsedSnippet {
+interface VisualSnippet extends _Snippet {
+	type: "visual";
 	trigger: string;
-	replacement: string;
-	options: Options;
-	flags: string;
-	priority?: number;
-	description?: string;
-	
-	constructor(raw: RawSnippet) {
-		// normalize regex triggers
-		const resolved: RawSnippet = { ...raw, flags: raw.flags ?? "" };
-		if (raw.trigger instanceof RegExp) {
-			resolved.options = `r${raw.options}`;
-			resolved.trigger = raw.trigger.source;
-			// regex trigger flags and snippet flags get merged 
-			resolved.flags = `${raw.trigger.flags}${resolved.flags}`;
-		}
-
-		// filter out invalid flags
-		const validFlags = [
-			// "d", // doesn't affect the search
-			// "g", // doesn't affect the pattern match and is almost certainly undesired behavior
-			"i",
-			"m",
-			"s",
-			"u",
-			"v",
-			// "y", // almost certainly undesired behavior
-		];
-		resolved.flags = Array.from(new Set(resolved.flags.split("")))
-			.filter(flag => validFlags.includes(flag))
-			.join(""); 
-
-		const parsed = {...resolved, options: Options.fromSource(resolved.options)};
-		Object.assign(this, parsed);
-	}
+	replacement: string | Fn<[selection: string], string | false>;
 }
 
-export interface Environment {
-	openSymbol: string,
-	closeSymbol: string
+interface RegexSnippet extends _Snippet {
+	type: "regex";
+	trigger: string;
+	replacement: string | Fn<[match: RegExpExecArray], string>;
+	flags?: string;
+}
+
+interface StringSnippet extends _Snippet {
+	type: "string";
+	trigger: string;
+	replacement: string | Fn<[match: string], string>;
+	flags?: string;
```

Comment:
Is this `flags` property for `StringSnippet`s ever accessed anywhere? (I'm pretty sure `flag`s are only used for regex snippets atm.)

I think it might actually make more sense to normalise the `trigger` for `RegexSnippet`s to be RegExps, instead of strings as we are currently doing. Then, we'd no longer need to define `const regex = new RegExp(`${trigger}$`, snippet.flags);` in `run_snippets.ts`, and we'd no longer need to store `flags` for regex snippets, since the flags would simply be contained inside the RegExp object.

From what I can tell, this would also simplify some of the parsing/normalisation logic used in `parse_snippets.ts`.

Reply:
i think the major concern for this is it amounts to a pretty significant breaking change; but yes, flags are only ever relevant for regex snippets.

maybe there could be a codemod to transform string regex snippets, or even automatically do so for snippets defined in the editor, to try to mitigate the effects though, coupled with a deprecation period recommending switching to true regex if a string regex snippet is detected. but i'd still imagine it'd be quite a mess.

Reply:
Ah, I was referring to storing the trigger as a RegExp internally; users could still use string regex snippets perfectly well, the plugin would just normalise their string triggers to RegExps (as opposed to the other way round — normalising RegExp triggers to strings — which is what the code currently does)

(Currently, RegExp triggers are converted to strings. Then those strings are converted _back into_ RegExps when the snippet gets executed in `run_snippets.ts`. Makes sense to simply normalise to the "final form" the triggers are used in imo)

In future, I imagine we could also add flags for string snippets, such as "i" for case insensitivity (#30). To implement this we could simply internally map such string snippets to regex snippets with the same trigger, but with the "i" flag enabled.

Reply:
makes sense; but triggers still need to be strings in order to substitute snippet variables either way, and not doing so would probably amount to a major breaking change, seeing how the historical use case of the snippet variables is precisely for substituting in regex snippets as far as i can tell.

Reply:
Yeah, the intention would be to store the triggers as RegExps with snippet variables already substituted in (I have no intention of making a breaking change!).

If you're still opposed though, I suggest at least removing the unused `flag` property on `StringSnippet`, and moving the `trigger` property from the three derived snippet interfaces to `_Snippet` so we're only defining it once.